gax-java/gax/src/main/java/com/google/api/gax/core/GaxProperties.java

```diff
@@ -113,4 +127,26 @@ static String getRuntimeVersion() {
     // with hyphens.
     return javaRuntimeInformation.replaceAll("[^0-9a-zA-Z_\\\\.]", "-");
   }
+
+  /**
+   * Returns the current library version as reported by {BUNDLE_VERSION_KEY} in library's
+   * META-INF/MANIFEST. This should only be used if MANIFEST file does not contain a widely
+   * recognized version declaration such as Specific-Version OR Implementation-Version declared in
+   * Manifest Specification
+   * https://docs.oracle.com/javase/8/docs/technotes/guides/jar/jar.html#Manifest_Specification,
+   * otherwise please use #getLibraryVersion
+   */
+  @VisibleForTesting
+  static Optional<String> getBundleVersion(Class<?> clazz) {
+    try {
+      File file = new File(clazz.getProtectionDomain().getCodeSource().getLocation().toURI());
+      try (JarFile jar = new JarFile(file.getPath())) {
+        Attributes attributes = jar.getManifest().getMainAttributes();
+        return Optional.ofNullable(attributes.getValue(BUNDLE_VERSION_KEY));
```

Comment:
I think we can inline BUNDLE_VERSION_KEY if it is not expected to be reused.

Reply:
done